forc-pkg/src/pkg.rs

```diff
@@ -1907,22 +1934,149 @@ pub fn build(plan: &BuildPlan, profile: &BuildProfile) -> anyhow::Result<(Compil
             namespace_map.insert(node, namespace.into());
         }
         json_abi.extend(compiled.json_abi);
+        json_abi_program
+            .types
+            .extend(compiled.json_abi_program.types);
+        json_abi_program
+            .functions
+            .extend(compiled.json_abi_program.functions);
         storage_slots.extend(compiled.storage_slots);
         bytecode = compiled.bytecode;
         tree_type = Some(compiled.tree_type);
         source_map.insert_dependency(manifest.dir());
     }
+
+    standardize_json_abi_types(&mut json_abi_program);
+
     let tree_type =
         tree_type.ok_or_else(|| anyhow!("build plan must contain at least one package"))?;
     let compiled = Compiled {
         bytecode,
         json_abi,
+        json_abi_program,
         storage_slots,
         tree_type,
     };
     Ok((compiled, source_map))
 }
 
+/// Standardize the JSON ABI data structure by eliminating duplicate types. This is an iterative
+/// process because every time two types are merged, new opportunities for more merging arise.
+fn standardize_json_abi_types(json_abi_program: &mut JsonABIProgram) {
```

Comment:
Good catch, I didn't think about this possibility... it makes me want to implement this in the `TypeEngine` 😄